compiler/rustc_mir_transform/src/lower_intrinsics.rs

```diff
@@ -194,7 +194,17 @@ impl<'tcx> MirPass<'tcx> for LowerIntrinsics {
                     }
                     sym::transmute => {
                         let dst_ty = destination.ty(local_decls, tcx).ty;
-                        if let Some(target) = *target {
+
+                        // The MIR validator is fine with `Call`s* where the `destination` is
+                        // something like `&!`, but not with `Assign`s to such a place, so we
+                        // have to check that the destination type doesn't create one.
```

Comment:
Note that I might also have interpreted things here in correctly -- I've this comment, and its related code approach, is completely gone from the current version of the PR.

Basically, `const_prop` was ending up failing in 

https://github.com/rust-lang/rust/blob/1db9c061d387a66ab16a90ec5a1b90adf216e2da/compiler/rustc_const_eval/src/interpret/validity.rs#L416-L421

as part of checking the place in an `Assignment`.

That's probably not the normal "the MIR validator", rather "validity checks that happen in MIR", which uses all the same root words but I think means something different.

Reply:
(This check in particular feels odd to me, since I thought that UCG was leaning towards under-a-reference validity being a safety invariant, not a validity one.)

Reply:
Possibly related: https://github.com/rust-lang/unsafe-code-guidelines/issues/77

My short summary is that references to uninhabited types _may_ be different from shallow vs deep validity. I'm not personally convinced of the optimization point, but frankly many of the arguments offered up in that thread don't strike me as particularly compelling.

In any case, the prudent thing is probably to reserve this case as UB just in case, especially if it isn't causing anyone problems for it to be UB currently.

Reply:
Oh, so this is the same underlying issue as that ICE, where ConstProp calls checks that the interpreter value is valid for a given type.

IMO it would be better to work around this quirk in ConstProp, rather than having lower_intrinsics go through some strange hoops that really shouldn't be necessary. This is clearly a ConstProp bug.

Reply:
I'll note that `lower_intrinsics` isn't doing this check in the current version of this PR -- this PR runs constprop on more transmutes than the none of them that it did before, but still doesn't try *all* transmutes exactly to workaround this issue.

So I think we can consider this thread resolved for this PR, and move any follow-up work to tweak validity checking to another issue.

Reply:
Ah, I didn't realize this was changed since I added the comment.